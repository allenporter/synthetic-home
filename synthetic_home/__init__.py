"""
.. include:: ../README.md
"""

__all__ = [
    "synthetic_home",
    "device_types",
    "exceptions",
    "registry",
    "tool",
]
