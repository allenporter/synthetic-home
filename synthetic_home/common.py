"""Common data model objects used across this library.

These are typically small objects reusable in different contexts.
"""

from dataclasses import dataclass

__all__ = ["DeviceInfo"]


@dataclass
class DeviceInfo:
    """Device model information."""

    model: str | None = None
    """The model name of the device e.g. 'Learning Thermostat'."""

    manufacturer: str | None = None
    """The manufacturer of the device e.g. 'Nest'."""

    sw_version: str | None = None
    """The firmware version string of the device e.g. '1.0.2'."""
