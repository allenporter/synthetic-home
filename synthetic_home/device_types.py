"""Data model for device type definitions.

These device devices and how they map to a set of entities. States can be set
on a device then translated into how that state affects entities.
"""

from functools import cache
from collections.abc import Generator
from dataclasses import dataclass, field
from importlib import resources
from importlib.resources.abc import Traversable
import logging
from typing import Any
import yaml

from mashumaro.codecs.yaml import yaml_decode
from mashumaro.exceptions import MissingField
from mashumaro import field_options, DataClassDictMixin
from mashumaro.types import SerializationStrategy

from .exceptions import SyntheticHomeError

__all__ = [
    "DeviceTypeRegistry",
    "DeviceType",
    "DeviceState",
    "EntityState",
    "EntityEntry",
    "load_device_type_registry",
]


_LOGGER = logging.getLogger(__name__)


DEVICE_TYPES_RESOURCE_PATH = resources.files().joinpath("registry")


class KeyedObjectListStrategy(SerializationStrategy):
    """A predefined entity state parser."""

    def __init__(self, object_strategy: SerializationStrategy) -> None:
        """Initialize KeyedObjectStrategy."""
        self._object_strategy = object_strategy

    def deserialize(self, value: Any) -> list[Any]:
        """Deserialize the object."""
        if not value:
            return []
        if isinstance(value, dict):
            return [
                self._object_strategy.deserialize((key, state))
                for key, state in value.items()
            ]
        raise ValueError(f"Expected 'dict' representing the object list, got: {value}")


@dataclass
class EntityEntry(DataClassDictMixin):
    """Defines an entity type.

    An entity is the lowest level object that maps to a behavior or trait of a device.
    """

    key: str
    """The entity description key"""

    attributes: dict[str, str | list[str]] = field(default_factory=dict)
    """Attributes supported by this entity."""


class EntityDictStrategy(SerializationStrategy):
    """A parser of the entity entry dict."""

    def deserialize(
        self, value: dict[str, dict[str, dict[str, str | list[str]]]]
    ) -> dict[str, list[EntityEntry]]:
        """Deserialize the object."""
        return {
            domain: [
                EntityEntry(key=key, attributes=attributes)
                for key, attributes in values.items()
            ]
            for domain, values in value.items()
        }


@dataclass
class EntityState(DataClassDictMixin):
    """Represents a pre-defined entity state."""

    domain: str
    """The domain of the entity."""

    key: str
    """The entity key identifying the entity."""

    state: str | dict[str, Any]
    """The values that make up the entity state."""

    @property
    def domain_key(self) -> str:
        """Return the full domain.key."""
        return f"{self.domain}.{self.key}"

    def merge(self, new: "EntityState") -> "EntityState":
        """Merge with an additional value."""
        merged_state: str | bool | dict[str, Any]
        if isinstance(new.state, dict):
            if isinstance(self.state, dict):
                merged_state = {
                    **self.state,
                    **new.state,
                }
            else:
                merged_state = {"state": self.state, **new.state}
        else:
            if isinstance(self.state, dict):
                merged_state = {
                    **self.state,
                    "state": new.state,
                }
            else:
                merged_state = new.state
        return EntityState(
            domain=self.domain,
            key=self.key,
            state=merged_state,
        )


def merge_entity_state_attributes(
    platform: str, entry: EntityEntry, entity_states: list[EntityState]
) -> EntityEntry:
    """Merge state values from an EntityEntry into a new EntityEntry."""
    for state in entity_states:
        if state.domain == platform and state.key == entry.key:
            if isinstance(state.state, dict):
                extras = state.state
            else:
                extras = {"state": state.state}
            return EntityEntry(
                key=entry.key,
                attributes={
                    **entry.attributes,
                    **extras,
                },
            )
    _LOGGER.debug("No state to merge: %s", entry)
    return entry


class EntityStateStrategy(SerializationStrategy):
    """A predefined entity state parser."""

    def deserialize(self, value: tuple[str, Any]) -> EntityState:
        """Deserialize the object."""
        key = value[0]
        state = value[1]
        parts = key.split(".")
        if len(parts) != 2:
            raise ValueError(f"Expected '<domain>.<entitiy-key>', got {key}")
        return EntityState(domain=parts[0], key=parts[1], state=state)


@dataclass
class DeviceState(DataClassDictMixin):
    """Represents a pre-defined device state.

    This is used to map a pre-defined devie state to the values that should be
    set on individual entities. These are typically "interesting" states of the
    device that can be enumerated during evaluation. For example, instead of
    explicitly specifying specific temperature values, a predefined state could
    implement "warm" and "cool".
    """

    name: str
    """An identifier that names this state."""

    entity_states: list[EntityState]
    """An identifier for this set of attributes used for labeling"""

    def merge(self, new_state: "DeviceState") -> "DeviceState":
        """Return a new DeviceState with merged entity states."""
        states = {entity.domain_key: entity for entity in self.entity_states}
        for entity in new_state.entity_states:
            if entity.domain_key in states:
                states[entity.domain_key] = states[entity.domain_key].merge(entity)
            else:
                states[entity.domain_key] = entity
        return DeviceState(name=self.name, entity_states=list(states.values()))


class DeviceStateStrategy(SerializationStrategy):
    """A predefined device state parser."""

    _child_strategy = KeyedObjectListStrategy(EntityStateStrategy())

    def deserialize(self, value: tuple[str, Any]) -> DeviceState:
        """Deserialize the object."""
        if not isinstance(value, tuple):
            raise ValueError(
                f"Expected 'tuple' representing the DeviceState object, got: {value}"
            )
        return DeviceState(
            name=value[0], entity_states=self._child_strategy.deserialize(value[1])
        )


@dataclass(frozen=True)
class DeviceType(DataClassDictMixin):
    """Defines of device type."""

    device_type: str
    """The identifier for the device e.g. 'smart-lock'"""

    desc: str
    """The human readable description of the device."""

    device_states: list[DeviceState] = field(
        metadata=field_options(
            serialization_strategy=KeyedObjectListStrategy(DeviceStateStrategy())
        ),
    )
    """A series of different attribute values that are most interesting to use during evaluation."""

    entities: dict[str, list[EntityEntry]] = field(
        metadata=field_options(serialization_strategy=EntityDictStrategy()),
        default_factory=dict,
    )
    """Entity platforms and their entity description keys"""

    @property
    def device_states_dict(self) -> dict[str, DeviceState]:
        """Get the predefined state by the specified key."""
        return {state.name: state for state in self.device_states}

    @property
    def entity_dict(self) -> dict[str, EntityEntry]:
        """Get a flat map of all entity entries."""
        return {
            f"{platform}.{entry.key}": entry
            for platform, entries in self.entities.items()
            for entry in entries
        }

    def __post_init__(self) -> None:
        """Validate the DeviceType."""
        entity_dict = self.entity_dict
        for device_state in self.device_states:
            for entity_state in device_state.entity_states:
                if entity_state.domain_key not in entity_dict:
                    raise ValueError(
                        f"Device '{self.device_type}' state '{device_state.name}' references "
                        f"invalid entity '{entity_state.domain_key}' not in {list(entity_dict.keys())}"
                    )


@dataclass
class DeviceTypeRegistry:
    """The registry of all DeviceType objects."""

    device_types: dict[str, DeviceType] = field(default_factory=dict)


def _read_device_types(
    device_types_path: Traversable,
) -> Generator[DeviceType, None, None]:
    """Read device types from the device type directory."""
    _LOGGER.debug("Loading device type registry from %s", device_types_path.absolute())  # type: ignore[attr-defined]

    for device_type_file in device_types_path.iterdir():
        if not device_type_file.name.endswith(".yaml"):
            continue
        try:
            with device_type_file.open("r") as f:
                content = f.read()
        except FileNotFoundError:
            raise SyntheticHomeError(
                f"Configuration file '{device_type_file}' does not exist"
            )

        try:
            device_type: DeviceType = yaml_decode(content, DeviceType)
        except MissingField as err:
            raise SyntheticHomeError(f"Unable to decode file {device_type_file}: {err}")
        except yaml.YAMLError as err:
            raise SyntheticHomeError(f"Unable to decode file {device_type_file}: {err}")
        if device_type_file.name != f"{device_type.device_type}.yaml":
            raise SyntheticHomeError(
                f"Device type '{device_type.device_type}' name does not match filename '{device_type_file.name}'"
            )

        yield device_type


@cache
def load_device_type_registry() -> DeviceTypeRegistry:
    """Load device types from the yaml configuration files."""
    device_types = {}
    for device_type in _read_device_types(DEVICE_TYPES_RESOURCE_PATH):
        if device_type.device_type in device_types:
            raise SyntheticHomeError(
                f"Device registry contains duplicate device type '{device_type.device_type}"
            )
        device_types[device_type.device_type] = device_type
    return DeviceTypeRegistry(device_types=device_types)
