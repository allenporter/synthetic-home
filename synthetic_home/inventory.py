"""Data model for the lower level inventory of a home, usable for fixtures and evaluations."""

import pathlib
from dataclasses import dataclass, field
import logging
import slugify
from typing import Any

import yaml
from mashumaro.mixins.yaml import DataClassYAMLMixin, EncodedData
from mashumaro.config import BaseConfig
from mashumaro.codecs.yaml import yaml_decode

from . import common
from .exceptions import SyntheticHomeError

__all__ = [
    "Inventory",
    "load_inventory",
    "Area",
    "Device",
    "Entity",
]

_LOGGER = logging.getLogger(__name__)

DEFAULT_SEPARATOR = "_"


def _custom_encoder(data: dict[str, Any]) -> EncodedData:
    return yaml.safe_dump(data, sort_keys=False, explicit_start=True)


@dataclass
class Area(DataClassYAMLMixin):
    """Represents an area in a home."""

    name: str
    """The human readable name of the area e.g. 'Living Room'."""

    id: str | None = None
    """The identifier of the area eg. 'living_room', unique within a Home."""

    floor: str | None = None
    """The human readable name of the floor."""

    def __post_init__(self) -> None:
        """Validate the area."""
        if self.id is None:
            self.id = slugify.slugify(self.name, separator=DEFAULT_SEPARATOR)

    class Config(BaseConfig):
        code_generation_options = ["TO_DICT_ADD_OMIT_NONE_FLAG"]
        sort_keys = False


@dataclass
class Device(DataClassYAMLMixin):
    """Represents a devie within a home."""

    name: str
    """The human readable name of the device e.g. 'Left Blind'."""

    id: str | None = None
    """The identifier of the area eg. 'living_room_left_blind', unique within a Home."""

    area: str | None = None
    """The area id of the home. e.g. 'living_room'"""

    info: common.DeviceInfo | None = None
    """Detailed model information about the device."""

    class Config(BaseConfig):
        code_generation_options = ["TO_DICT_ADD_OMIT_NONE_FLAG"]
        sort_keys = False

    def __post_init__(self) -> None:
        """Validate the device."""
        if self.id is None:
            self.id = slugify.slugify(self.name, separator=DEFAULT_SEPARATOR)


@dataclass
class Entity(DataClassYAMLMixin):
    """Represents an entity within a home."""

    name: str | None = None
    """The human readable name of the entity e.g. 'Outside rain sensor'. or omitted to use device naming."""

    id: str | None = None
    """The identifier of the entity e.g. 'sensor.rain_sensor_intensity', unique within a Home."""

    area: str | None = None
    """The area id within the home e.g. 'living_room'."""

    device: str | None = None
    """The device id within the home e.g. 'living_room_left_blind'."""

    state: str | None = None
    """The current state value for the entity."""

    attributes: dict[str, str | int | float | bool | list[str]] | None = None
    """The current state attributes for the entity."""

    def __post_init__(self) -> None:
        """Validate the device."""
        if self.id is None:
            raise SyntheticHomeError("Entity {self.name} had no value for 'id'")

    class Config(BaseConfig):
        code_generation_options = ["TO_DICT_ADD_OMIT_NONE_FLAG"]
        sort_keys = False


@dataclass
class Inventory(DataClassYAMLMixin):
    """A fixture defines the entire definition of the home."""

    language: str | None = None
    """The default system language."""

    areas: list[Area] = field(default_factory=list)
    """The list of areas within the home."""

    devices: list[Device] = field(default_factory=list)
    """The list of devices within a home, which may be associated with areas."""

    entities: list[Entity] = field(default_factory=list)
    """The list of entities within a home, which may be associated with areas and devices."""

    def yaml(self) -> str:
        """Render the inventory as yaml."""
        return str(self.to_yaml(omit_none=True, encoder=_custom_encoder))

    @property
    def floors(self) -> set[str]:
        """Return the set of floors across all areas."""
        return {area.floor for area in self.areas if area.floor is not None}

    def device_dict(self) -> dict[str, Device]:
        """Dictionary of devices by device id."""
        return {device.id: device for device in self.devices if device.id is not None}

    def area_dict(self) -> dict[str, Area]:
        """Dictionary of areas by area id."""
        return {area.id: area for area in self.areas if area.id is not None}

    class Config(BaseConfig):
        code_generation_options = ["TO_DICT_ADD_OMIT_NONE_FLAG"]
        sort_keys = False


def read_config_content(config_file: pathlib.Path) -> str:
    """Create configuration file content, exposed for patching."""
    with config_file.open("r") as f:
        return f.read()


def decode_inventory(content: str) -> Inventory:
    """Load synthetic home configuration from disk."""
    return yaml_decode(content, Inventory)


def load_inventory(config_file: pathlib.Path) -> Inventory:
    """Load synthetic home configuration from disk."""
    try:
        content = read_config_content(config_file)
    except FileNotFoundError:
        raise SyntheticHomeError(f"Configuration file '{config_file}' does not exist")
    try:
        return decode_inventory(content)
    except ValueError as err:
        raise SyntheticHomeError(f"Could not parse config file '{config_file}': {err}")
