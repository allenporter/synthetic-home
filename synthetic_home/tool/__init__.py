"""Synthetic home command line tools."""

__all__ = [
    "create_inventory",
    "export_inventory",
    "list_device_types",
]
