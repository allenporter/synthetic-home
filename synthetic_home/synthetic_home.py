"""Data model for home assistant synthetic home."""

from dataclasses import dataclass, field
import pathlib
import logging
import slugify
from typing import Any

from mashumaro.codecs.yaml import yaml_decode

from synthetic_home.exceptions import SyntheticHomeError
from synthetic_home.device_types import (
    DeviceTypeRegistry,
    DeviceStateStrategy,
    DeviceState,
    EntityEntry,
    merge_entity_state_attributes,
)
from . import common, inventory
from .inventory import DEFAULT_SEPARATOR
from .device_types import load_device_type_registry

__all__ = [
    "SyntheticHome",
    "Device",
    "build_device_state",
    "load_synthetic_home",
    "read_config_content",
]


_LOGGER = logging.getLogger(__name__)


@dataclass
class Device:
    """A synthetic device."""

    name: str
    """A human readable name for the device."""

    device_type: str | None = None
    """The type of the device in the device registry that determines how it maps to entities."""

    device_info: common.DeviceInfo | None = None
    """Device make and model information."""

    device_state: str | DeviceState | None = None
    """A list of pre-canned RestorableStateAttributes specified by the key.

    These are used for restoring a device into a specific state supported by the
    device type. This is used to use a label rather than specifying low level
    entity details. This is an alternative to specifying low level attributes above.

    Restorable attributes overwrite normal attributes since they can be reloaded
    at runtime.
    """

    entity_entries: dict[str, list[EntityEntry]] = field(default_factory=dict)
    """The validated set of entity entries"""

    def merge(
        self,
        device_state: DeviceState | None = None,
        entity_entries: dict[str, list[EntityEntry]] | None = None,
    ) -> "Device":
        """Merge the existing device with a new device state."""
        return Device(
            name=self.name,
            device_type=self.device_type,
            device_info=self.device_info,
            device_state=device_state or self.device_state,
            entity_entries=entity_entries or self.entity_entries,
        )


def build_device_state(device: Device, registry: DeviceTypeRegistry) -> Device:
    """Validate the device and return a new instance."""
    if (device_type := registry.device_types.get(device.device_type or "")) is None:
        raise SyntheticHomeError(
            f"Device {device} has device_type {device.device_type} not found in: {registry.device_types}"
        )

    if device.device_state is not None:
        # Lookup a device state and merge it into the current state
        if (
            isinstance(device.device_state, str)
            and device.device_state not in device_type.device_states_dict
        ):
            raise SyntheticHomeError(
                f"Device {device} has state '{device.device_state}' not in: {device_type.device_states_dict}"
            )
        if isinstance(device.device_state, dict):
            _LOGGER.debug(
                "Parsing device state from dictionary: %s", device.device_state
            )
            strategy = DeviceStateStrategy()
            device = device.merge(
                device_state=strategy.deserialize(("custom", device.device_state))
            )
            _LOGGER.debug("Parsed=%s", device)

    device_state: DeviceState | None = None
    _LOGGER.debug("Checking device state: %s", device.device_state)
    if (
        device.device_state is None or isinstance(device.device_state, DeviceState)
    ) and device_type.device_states:
        # Pick the first device state as the default
        device_state = device_type.device_states[0]
        if isinstance(device.device_state, DeviceState):
            device_state = device_state.merge(device.device_state)
    elif device.device_state is not None and isinstance(device.device_state, str):
        device_state = device_type.device_states_dict[device.device_state]
    else:
        raise SyntheticHomeError(f"Device did not declare a device state: {device}")

    _LOGGER.debug("Merging entity attributes for device state: %s", device_state)
    entity_entries = {
        platform: [
            merge_entity_state_attributes(
                platform, entity_entry, device_state.entity_states
            )
            for entity_entry in entity_entries
        ]
        for platform, entity_entries in device_type.entities.items()
    }
    return device.merge(device_state=device_state, entity_entries=entity_entries)


@dataclass
class SyntheticHome:
    """Data about a synthetic home."""

    # Devices by area
    devices: dict[str, list[Device]] = field(default_factory=dict)

    # Services for the home not for a specific area
    services: list[Device] = field(default_factory=list)

    # Device types supported by the home.
    device_type_registry: DeviceTypeRegistry | None = None

    def __post_init__(self) -> None:
        """Build the complete device state."""
        if self.device_type_registry is None:
            self.device_type_registry = load_device_type_registry()
        self.devices = {
            key: [
                build_device_state(device, self.device_type_registry)
                for device in devices
            ]
            for key, devices in self.devices.items()
        }
        self.services = [
            build_device_state(device, self.device_type_registry)
            for device in self.services
        ]


def read_config_content(config_file: pathlib.Path) -> str:
    """Create configuration file content, exposed for patching."""
    with config_file.open("r") as f:
        return f.read()


def load_synthetic_home(config_file: pathlib.Path) -> SyntheticHome:
    """Load synthetic home configuration from disk."""
    try:
        content = read_config_content(config_file)
    except FileNotFoundError:
        raise SyntheticHomeError(f"Configuration file '{config_file}' does not exist")
    try:
        return yaml_decode(content, SyntheticHome)
    except ValueError as err:
        raise SyntheticHomeError(f"Could not parse config file '{config_file}': {err}")


def yaml_state_value(v: Any) -> Any:
    """Convert a entity state value to yaml."""
    if isinstance(v, bool) or isinstance(v, float) or isinstance(v, list):
        return v
    return str(v)


def build_entities(area_id: str | None, device_entry: Device) -> list[inventory.Entity]:
    """Build the set of entities for the device entry."""
    entities = []
    device_name = device_entry.name.replace("_", " ").title()
    device_id = slugify.slugify(device_entry.name, separator=DEFAULT_SEPARATOR)
    for platform, entity_entries in device_entry.entity_entries.items():
        for entity_entry in entity_entries:
            # Each entity in this platform needs a unique name, but
            # if the key is in the name it's the primary to avoid "Motion motion"
            entity_name = device_name
            if platform == "sensor" or (
                platform == "binary_sensor"
                and entity_entry.key not in device_entry.name.lower()
            ):
                entity_name = f"{device_name} {entity_entry.key.capitalize()}"
            entity_id = f"{platform}.{slugify.slugify(entity_name, separator=DEFAULT_SEPARATOR)}"
            entity = inventory.Entity(
                name=entity_name,
                id=entity_id,
                device=device_id,
            )
            if area_id:
                entity.area = area_id
            attributes: dict[str, str | list[str] | int | float] = {}
            if entity_entry.attributes:
                attributes.update(entity_entry.attributes)
            state = attributes.pop("state", None)
            if state is not None:
                entity.state = yaml_state_value(state)
            if attributes:
                entity.attributes = attributes
            entities.append(entity)
    return entities


def build_inventory(home: SyntheticHome) -> inventory.Inventory:
    """Build a home inventory from the synthetic home definition.

    This is a flattened set of areas, entities, and devices.
    """

    inv = inventory.Inventory()
    pairs: list[tuple[str | None, list[Device]]] = [*home.devices.items()]
    if home.services:
        pairs.append((None, home.services))

    entities = []
    for area_name, devices in pairs:
        if area_name:
            area_id = slugify.slugify(area_name, separator=DEFAULT_SEPARATOR)
            inv.areas.append(inventory.Area(name=area_name, id=area_id))
        else:
            area_id = None

        for device_entry in devices:
            # Make computer generated device names more friendly
            device_name = device_entry.name.replace("_", " ").title()
            device_id = slugify.slugify(device_entry.name, separator=DEFAULT_SEPARATOR)
            device = inventory.Device(
                name=device_name,
                id=device_id,
                info=device_entry.device_info,
            )
            if area_id:
                device.area = area_id
            inv.devices.append(device)
            entities.extend(build_entities(area_id, device_entry))
    if entities:
        inv.entities = entities
    return inv
