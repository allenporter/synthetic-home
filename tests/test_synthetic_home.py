"""Test for synthetic_home."""

import pathlib

import pytest
from syrupy import SnapshotAssertion

from synthetic_home import synthetic_home, device_types, inventory


TEST_HOMES = pathlib.Path("tests/homes")
TEST_FIXTURES = pathlib.Path("tests/fixtures")
HOME1 = TEST_HOMES / "home1.yaml"


def test_load_synthetic_home() -> None:
    """Test loading the device type registry."""

    home = synthetic_home.load_synthetic_home(HOME1)
    assert home
    area = home.devices.get("Backyard")
    assert area
    assert len(area) == 1
    device = area[0]
    assert device.name == "Outdoor Camera"
    assert device.device_info
    assert device.device_info.model == "Spotlight Cam Battery"
    assert device.device_info.manufacturer == "Ring"
    assert device.device_info.sw_version == "2.4.1"


@pytest.mark.parametrize(
    ("home_filename"),
    list(TEST_HOMES.glob("*.yaml")),
    ids=[str(filename) for filename in TEST_HOMES.glob("*.yaml")],
)
def test_inventory(home_filename: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test converting the home into an inventory yaml file."""

    # Load the inventory fixture
    inventory_filename = TEST_FIXTURES / home_filename.name
    inventory_content = inventory_filename.read_text().strip()

    # Generate the inventory fixture from the synthetic home device state
    home = synthetic_home.load_synthetic_home(home_filename)
    inventory = synthetic_home.build_inventory(home)
    generated_yaml = inventory.yaml().strip()

    assert generated_yaml.split("\n") == inventory_content.split("\n")


@pytest.mark.parametrize(
    ("filename"),
    list(TEST_HOMES.glob("*.yaml")),
    ids=[str(filename) for filename in TEST_HOMES.glob("*.yaml")],
)
def test_device_state_to_entity_state(
    filename: pathlib.Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading the device states and how they correspond to entity states."""
    home = synthetic_home.load_synthetic_home(filename)
    device_registry = device_types.load_device_type_registry()
    for devices in home.devices.values():
        for device in devices:
            assert device.device_type
            device_type = device_registry.device_types[device.device_type]
            assert device_type
            for device_state in device_type.device_states:
                assert device_state
                device.device_state = device_state
                device = synthetic_home.build_device_state(device, device_registry)

                inv = inventory.Inventory()
                inv.entities = synthetic_home.build_entities(None, device)
                assert inv.yaml() == snapshot
