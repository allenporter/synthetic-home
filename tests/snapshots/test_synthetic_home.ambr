# serializer version: 1
# name: test_inventory
  '''
  ---
  areas:
  - name: Backyard
    id: backyard
  devices:
  - name: Outdoor Camera
    id: backyard_outdoor_camera
    area: backyard
    info:
      model: Spotlight Cam Battery
      manufacturer: Ring
      sw_version: 2.4.1
  entities: []
  
  '''
# ---
